src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java

```diff
@@ -0,0 +1,221 @@
+        this.internal = new JDKDirectoryWatch(path, exec, internalEventHandler) {
+
+            // Override to ensure that this watch relativizes events wrt
+            // `rootPath` (instead of `path`, as is the default behavior)
+            @Override
+            public WatchEvent relativize(WatchEvent event) {
+                return new WatchEvent(event.getKind(), rootPath,
+                    rootPath.relativize(event.calculateFullPath()));
+            }
+
+            // Override to ensure that this watch translates JDK events using
+            // `rootPath` (instead of `path`, as is the default behavior).
+            // Events returned by this method do not need to be relativized.
+            @Override
+            protected WatchEvent translate(java.nio.file.WatchEvent<?> jdkEvent) {
+                var kind = translate(jdkEvent.kind());
+
+                Path relativePath = null;
+                if (kind != WatchEvent.Kind.OVERFLOW) {
+                    var child = (Path) jdkEvent.context();
+                    if (child != null) {
+                        relativePath = relativePathParent.resolve(child);
+                    }
+                }
+
+                var event = new WatchEvent(kind, rootPath, relativePath);
+                logger.trace("Translated: {} to {}", jdkEvent, event);
+                return event;
+            }
+        };
+    }
+
+    /**
+     * Event handler that updates the child watches according to the following
+     * rules: (a) when an overflow happens, it's propagated to each existing
+     * child watch; (b) when a subdirectory creation happens, a new child watch
+     * is opened for that subdirectory; (c) when a subdirectory deletion
+     * happens, an existing child watch is closed for that subdirectory.
+     */
+    private class ChildWatchesUpdater implements BiConsumer<EventHandlingWatch, WatchEvent> {
+        @Override
+        public void accept(EventHandlingWatch watch, WatchEvent event) {
+            switch (event.getKind()) {
+                case OVERFLOW: acceptOverflow(); break;
+                case CREATED: acceptCreated(event.calculateFullPath()); break;
+                case DELETED: acceptDeleted(event.calculateFullPath()); break;
+                case MODIFIED: break;
+            }
+        }
+
+        private void acceptOverflow() {
+            for (var childWatch : childWatches.values()) {
+                reportOverflowTo(childWatch);
+            }
+        }
+
+        private void acceptCreated(Path fullPath) {
+            if (Files.isDirectory(fullPath)) {
+                var childWatch = openChildWatch(fullPath);
+                // Events in the newly created directory might have been missed
+                // between its creation and setting up its watch. So, generate
+                // an `OVERFLOW` event for the watch.
+                reportOverflowTo(childWatch);
```

Comment:
With this strategy the semantic of this recursive watcher depend on the overflow handling mode. I would prefer to keep that separate. Not to mention we generate overflow events a lot more often. I get the proposal: well, this is like an overflow, so it's up to the user to deal with that?

Are we sure that the modification time tracker knows how to deal with an overflow right at (or after) its creation?